### PR TITLE
Finetune on Iceberg Data

### PR DESCRIPTION
**Pull request by IsaacWarren**
## Changes included in this PR

Adds a new example.

## Testing strategy
Tested on my laptop.

## User facing changes
New example

## Checklist
- [ ] Pipelines passed before requesting review. To run CI you must include `[run CI]` in your commit message.
- [x] I am familiar with the [Contributing Guide](https://github.com/bodo-ai/Bodo/blob/main/CONTRIBUTING.md) 
- [x] I have installed + ran pre-commit hooks.

### CODE REVIEW

File: examples/ML/llama_finetune_iceberg_data.py
Comment (IsaacWarren) [inline examples/ML/llama_finetune_iceberg_data.py:1]: @strangeloopcanon @scott-routledge2 Can you guys take a look at this and see if it makes sense to you before I clean up/start on the blog

Comment (strangeloopcanon) [reply examples/ML/llama_finetune_iceberg_data.py:1]: Few thoughts:
- attention mask didn't move to device? It's a CPU going into GPU 
- padding should prob use the tokenizer's id, not 0, otherwise it'll corrup. I'd also add an EOS to output_tokenizerd
- maybe compute loss average with the true batch size per step
- maybe also do model.config.use_cache=False for training
- load the tokenizer once and just pass as argument in the if __name__ == "__main__" block
- It iterates idxs once to find max_len, then iterate again to build tensors. For large batches I'd compute max_len in a single pass
-  save is with torch.distributed.checkpoint.save but then immediately call base_model.save_pretrained with rank 0, prob worth just using torch evwhere?
- oh and not sure this was intentional or not but the lora target now hits all linear projections, just note its quite aggressive sinceits training most of the model. maybe just do attention projections?
- i'd gradient clip! before optimizer step
- Minor but I'd also split checkpoint dirs for adapter and checkpoint

File: examples/ML/llama_finetune_iceberg_data.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        input_ids = torch.tensor(row["input_tokenized"])
+        output_ids = torch.tensor(row["output_tokenized"])
+        example = torch.cat([input_ids, output_ids])
+        labels = torch.cat([torch.full_like(input_ids, -100), output_ids])
+        attention_mask = torch.ones_like(example)
+
+        return example, labels, attention_mask
+
+    def __getitems__(self, idxs):
+        input_ids = []
```
Comment (scott-routledge2) [inline examples/ML/llama_finetune_iceberg_data.py]: Could we just do `tokenizer(list_batch_input_output, padding="max_len")` here/add a collate_fn? I guess the tradeoff would be Tokenizing an entire dataset up front might not be practical for large data but could potentially save some compute if the entire result can fit in memory.

File: examples/ML/llama_finetune_iceberg_data.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        
+        torch.distributed.checkpoint.save(
+            {"model_state_dict": adapter_state_dict}, # Save only the adapter
+            checkpoint_id=CHECKPOINT_DIR
+        )
+        
+    # Save in peft-preferred format on rank 0
+    if pytorch_rank == 0:
+       base_model.save_pretrained(CHECKPOINT_DIR) # Saves adapter_config.json etc.
+
```
Comment (scott-routledge2) [inline examples/ML/llama_finetune_iceberg_data.py:243]: Do some evaluation/inference at the end? I'm not sure if a validation set makes sense here but it would be good to confirm the model can at least perform pretty well on the training data (memorize some examples)

Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py:243]: The problem is the training data is just random text so it won't actually improve

File: examples/ML/llama_finetune_iceberg_data.py (hunk 1 above)
Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py]: Yeah I was trying to use our tokenize function to show it off but it might not be the best fit

File: examples/ML/llama_finetune_iceberg_data.py (hunk 2 above)
Comment (scott-routledge2) [reply examples/ML/llama_finetune_iceberg_data.py:243]: Oh makes sense. I think it should still be able memorize it's training data if it's small enough? more of a test to check the config is correct but not necessary

Comment (scott-routledge2) [reply examples/ML/llama_finetune_iceberg_data.py:243]: Also did we want to add the data gen script to the example?

Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py:243]: The data is public, do you think we still should?

File: examples/ML/llama_finetune_iceberg_data.py (hunk 1 above)
Comment (scott-routledge2) [reply examples/ML/llama_finetune_iceberg_data.py]: Maybe we could add some version of this to `datasets.py` called something like `TokensDataset`

File: examples/ML/llama_finetune_iceberg_data.py
Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py:1]: Thanks! I've applied most of these, just a couple questions on a few

> It iterates idxs once to find max_len, then iterate again to build tensors. For large batches I'd compute max_len in a single pass

Not sure what this means, you have to iterate twice to find the max first if you need the max in the loop

> save is with torch.distributed.checkpoint.save but then immediately call base_model.save_pretrained with rank 0, prob worth just using torch evwhere?

It's two different formats, one is the torch one every epoch, the other is peft's format so it can easily be applied as a lora

> h and not sure this was intentional or not but the lora target now hits all linear projections, just note its quite aggressive sinceits training most of the model. maybe just do attention projections?

It's training about .5% of the parameters right now but that could make it even faster, I'll give it a shot!

> Minor but I'd also split checkpoint dirs for adapter and checkpoint

It only saves the adapter, just in different formats. I just wasn't sure how distributed model would interact with PEFT's saving so I only did it when training was done.

File: examples/ML/llama_finetune_iceberg_data.py (hunk 2 above)
Comment (scott-routledge2) [reply examples/ML/llama_finetune_iceberg_data.py:243]: Oh didn't realize it was a public. It's probably not necessary then, unless maybe we want to show writing an iceberg table with Bodo for the blog?

Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py:243]: > Oh didn't realize it was a public. It's probably not necessary then, unless maybe we want to show writing an iceberg table with Bodo for the blog? 

I did it with pyicberg+pyarrow 😂

File: examples/ML/llama_finetune_iceberg_data.py
Comment (strangeloopcanon) [reply examples/ML/llama_finetune_iceberg_data.py:1]: Ah misunderstood the intent re 2 saves, maybe a comment in the code to explain.

Re idx point it need two phases (find max, then pad), but you don't need two passes over the dataframe was my thought. In __getitems__, you iterate idxs twice and call df.iloc[idx] + tensor construction twice for the same indices. Prob academic for batch size 2 though, will only matter if it's >> higher.

File: examples/ML/llama_finetune_iceberg_data.py (hunk 1 above)
Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py]: Maybe, I just don't see a way to make it generic enough?

File: examples/ML/llama_finetune_iceberg_data.py
Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py:1]: Gotcha, I reworked it to use the tokenizer so shouldn't be a problem anymore

File: examples/ML/llama_finetune_iceberg_data.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
+import bodo.pandas as pd
+
+# Use pandas for timestamp manipulations
+import pandas
```
Comment (ehsantn) [inline examples/ML/llama_finetune_iceberg_data.py]: Seems unnecessary now? Is there a gap for us to fill?
```suggestion
```

Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py]: Good catch, I was using a timedelta to get an offset from the current date but realized that wouldn't work since the dates in the table are static so just used a set cutoff date instead

Comment (scott-routledge2) [reply examples/ML/llama_finetune_iceberg_data.py]: Just tried `pd.Timestamp.now()` and it doesn't work, since we wrap everything in `fallback_wrapper` by default I think.

Comment (ehsantn) [reply examples/ML/llama_finetune_iceberg_data.py]: Let's open issues for these and fix them soon since seem important.

Comment (IsaacWarren) [reply examples/ML/llama_finetune_iceberg_data.py]: https://bodo.atlassian.net/browse/BSE-5179?atlOrigin=eyJpIjoiMzBmMDlkMjViMDc2NDMxOWFmNDJkMDc1MGQ2NTljMDQiLCJwIjoiaiJ9
